### PR TITLE
Fixed resource_aci_l3extout idempotency issue and relationship attribute import issues

### PR DESCRIPTION
**Pull request by sajagana**
Note:

Fixed resource_aci_l3extout idempotency issue and relationship attribute import issues.

### CODE REVIEW

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

File: aci/data_source_aci_l3extout.go [hunk 1]
```diff
@@ -28,26 +29,62 @@ func dataSourceAciL3Outside() *schema.Resource {
 			},
 
 			"enforce_rtctrl": {
-				Type:     schema.TypeList,
+				Type:     schema.TypeString,
```
[the remark this answers is not in the document]
Comment (akinross) [reply aci/data_source_aci_l3extout.go:32]: You are correct this should be list will change this

File: aci/data_source_aci_l3extout.go [hunk 2]
```diff
@@ -36,18 +37,57 @@ func dataSourceAciL3Outside() *schema.Resource {
...
+			"relation_l3ext_rs_dampening_pol": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"tn_rtctrl_profile_name": {
+							Type:       schema.TypeString,
+							Optional:   true,
+							Computed:   true,
+							Deprecated: "Use tn_rtctrl_profile_dn instead of tn_rtctrl_profile_name",
```
Comment (lhercot) [inline aci/data_source_aci_l3extout.go:54]: I don't think we need the deprecation in the datasource

File: website/docs/d/l3_outside.html.markdown [hunk 3]
```diff
@@ -10,25 +10,41 @@ description: |-
...
+
 ## Example Usage ##
 
 ```hcl
-data "aci_l3_outside" "dev_l3_out" {
-  tenant_dn  = aci_tenant.dev_tenant.id 
-  name       = "foo_l3_out"
+data "aci_l3_outside" "foo_l3_outside" {
+  tenant_dn      = aci_tenant.terraform_tenant.id
+  name           = "foo_l3_outside"
```
Comment (lhercot) [inline website/docs/d/l3_outside.html.markdown:27]: Why adding spaces here?

File: website/docs/d/l3_outside.html.markdown [hunk 4]
```diff
@@ -10,25 +10,41 @@ description: |-
...
-* `annotation` - (Optional) annotation for object l3 outside.
-* `enforce_rtctrl` - (Optional) enforce route control type
-* `name_alias` - (Optional) name_alias for object l3 outside.
-* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+* `description`- (Optional) Description of the L3 Outside object.
+* `annotation` - (Optional) Annotation of the L3 Outside object.
+* `enforce_rtctrl` - (Optional) Enforce route control type of the L3 Outside object. 
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object.
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Cardinality - N_TO_M. Type - Block.
```
Comment (lhercot) [inline website/docs/d/l3_outside.html.markdown:43]: Relation to Route Control Profile for Dampening Policies (class rtctrlProfile).

File: website/docs/r/l3_outside.html.markdown [hunk 5]
```diff
@@ -11,35 +11,63 @@ description: |-
...
-- `relation_l3ext_rs_l3_dom_att` - (Optional) Relation to class extnwDomP. Cardinality - N_TO_ONE. Type - String.
+* `tenant_dn` - (Required) Distinguished name of the parent Tenant object.
+* `name` - (Required) Name of the L3 Outside object.
+* `description`- (Optional) Description of the L3 Outside object.
+* `annotation` - (Optional) Annotation of the L3 Outside object.
+* `enforce_rtctrl` - (Optional) Enforce route control type. Allowed values are "import" and "export". Default is "export". Type - String.
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object. Allowed values are "CS0", "CS1", "AF11", "AF12", "AF13", "CS2", "AF21", "AF22", "AF23", "CS3", "AF31", "AF32", "AF33", "CS4", "AF41", "AF42", "AF43", "CS5", "VA", "EF", "CS6", "CS7" and "unspecified". Default is "unspecified".
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Can't configure multiple Dampening Policies for the same address-family. Cardinality - N_TO_M. Type - Block.
```
Comment (lhercot) [inline website/docs/r/l3_outside.html.markdown:63]: Relation to Route Control Profile for Dampening Policies (class rtctrlProfile).

File: website/docs/r/l3_outside.html.markdown [hunk 6]
```diff
@@ -11,35 +11,63 @@ description: |-
...
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies. Allowed values are "ipv4-ucast" and "ipv6-ucast". Default is "ipv4-ucast".
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx. Target VRF object should belong to the parent tenant or be a shared object. Cardinality - N_TO_ONE. Type - String.
+* `relation_l3ext_rs_out_to_bd_public_subnet_holder` - (Optional) Relation to class fvBDPublicSubnetHolder. Create operation not allowed to the MO: l3extRsOutToBDPublicSubnetHolder. Cardinality - N_TO_M. Type - Set of String.
+* `relation_l3ext_rs_interleak_pol` - (Optional) Relation to class rtctrlProfile. Interleak Policy object should belong to the parent tenant or be a shared object. Cardinality - N_TO_ONE. Type - String.
+* `relation_l3ext_rs_l3_dom_att` - (Optional) Relation to class extnwDomP. Cardinality - N_TO_ONE. Type - String.
```
Comment (lhercot) [inline website/docs/r/l3_outside.html.markdown:70]: Relation to a L3 Domain (class extnwDomP)

File: website/docs/d/l3_outside.html.markdown [hunk 7]
```diff
@@ -10,25 +10,41 @@ description: |-
...
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies.
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx.
+* `relation_l3ext_rs_out_to_bd_public_subnet_holder` - (Optional) Relation to class fvBDPublicSubnetHolder.
+* `relation_l3ext_rs_interleak_pol` - (Optional) Relation to class rtctrlProfile.
+* `relation_l3ext_rs_l3_dom_att` - (Optional) Relation to class extnwDomP.
```
Comment (lhercot) [inline website/docs/d/l3_outside.html.markdown:50]: Relation to a L3 Domain (class extnwDomP)

File: website/docs/d/l3_outside.html.markdown [hunk 8]
```diff
@@ -10,25 +10,41 @@ description: |-
...
+* `enforce_rtctrl` - (Optional) Enforce route control type of the L3 Outside object. 
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object.
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Cardinality - N_TO_M. Type - Block.
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies.
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx.
+* `relation_l3ext_rs_out_to_bd_public_subnet_holder` - (Optional) Relation to class fvBDPublicSubnetHolder.
+* `relation_l3ext_rs_interleak_pol` - (Optional) Relation to class rtctrlProfile.
```
Comment (lhercot) [inline website/docs/d/l3_outside.html.markdown:49]: Relation to Route Profile for Interleak (class rtctrlProfile)

File: website/docs/r/l3_outside.html.markdown [hunk 9]
```diff
@@ -11,35 +11,63 @@ description: |-
...
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Can't configure multiple Dampening Policies for the same address-family. Cardinality - N_TO_M. Type - Block.
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies. Allowed values are "ipv4-ucast" and "ipv6-ucast". Default is "ipv4-ucast".
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx. Target VRF object should belong to the parent tenant or be a shared object. Cardinality - N_TO_ONE. Type - String.
+* `relation_l3ext_rs_out_to_bd_public_subnet_holder` - (Optional) Relation to class fvBDPublicSubnetHolder. Create operation not allowed to the MO: l3extRsOutToBDPublicSubnetHolder. Cardinality - N_TO_M. Type - Set of String.
+* `relation_l3ext_rs_interleak_pol` - (Optional) Relation to class rtctrlProfile. Interleak Policy object should belong to the parent tenant or be a shared object. Cardinality - N_TO_ONE. Type - String.
```
Comment (lhercot) [inline website/docs/r/l3_outside.html.markdown:69]: Relation to Route Profile for Interleak (class rtctrlProfile)

File: website/docs/r/l3_outside.html.markdown [hunk 10]
```diff
@@ -11,35 +11,63 @@ description: |-
...
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object. Allowed values are "CS0", "CS1", "AF11", "AF12", "AF13", "CS2", "AF21", "AF22", "AF23", "CS3", "AF31", "AF32", "AF33", "CS4", "AF41", "AF42", "AF43", "CS5", "VA", "EF", "CS6", "CS7" and "unspecified". Default is "unspecified".
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Can't configure multiple Dampening Policies for the same address-family. Cardinality - N_TO_M. Type - Block.
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies. Allowed values are "ipv4-ucast" and "ipv6-ucast". Default is "ipv4-ucast".
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx. Target VRF object should belong to the parent tenant or be a shared object. Cardinality - N_TO_ONE. Type - String.
```
Comment (lhercot) [inline website/docs/r/l3_outside.html.markdown:67]: Relation to VRF (class fvCtx)

File: website/docs/d/l3_outside.html.markdown [hunk 11]
```diff
@@ -10,25 +10,41 @@ description: |-
...
+* `description`- (Optional) Description of the L3 Outside object.
+* `annotation` - (Optional) Annotation of the L3 Outside object.
+* `enforce_rtctrl` - (Optional) Enforce route control type of the L3 Outside object. 
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object.
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to class rtctrlProfile. Cardinality - N_TO_M. Type - Block.
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies.
+* `relation_l3ext_rs_ectx` - (Optional) Relation to class fvCtx.
```
Comment (lhercot) [inline website/docs/d/l3_outside.html.markdown:47]: Relation to VRF (class fvCtx)

File: aci/data_source_aci_l3extout.go [hunk 12]
```diff
@@ -36,18 +37,57 @@ func dataSourceAciL3Outside() *schema.Resource {
...
+						},
+					},
+				},
+			},
+			"relation_l3ext_rs_ectx": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+			"relation_l3ext_rs_out_to_bd_public_subnet_holder": &schema.Schema{
```
Comment (lhercot) [inline aci/data_source_aci_l3extout.go:74]: This is not configurable so I think we should not add it.

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — DISMISSED** [review]
LGTM

File: aci/resource_aci_l3extout.go [hunk 13]
```diff
@@ -28,7 +28,7 @@ func resourceAciL3Outside() *schema.Resource {
 
 		SchemaVersion: 1,
 
-		Schema: AppendBaseAttrSchema(map[string]*schema.Schema{
+		Schema: AppendBaseAttrSchema(AppendNameAliasAttrSchema(map[string]*schema.Schema{
```
Comment (akinross) [inline aci/resource_aci_l3extout.go:31]: should we add mplsEnabled attribute as shown in model-doc?

File: aci/resource_aci_l3extout.go [hunk 14]
```diff
@@ -90,45 +85,51 @@ func resourceAciL3Outside() *schema.Resource {
 					"unspecified",
 				}, false),
 			},
-
+			// Relation to Route Control for Dampening
 			"relation_l3ext_rs_dampening_pol": &schema.Schema{
```
Comment (akinross) [inline aci/resource_aci_l3extout.go:89]: should we check adding all other missing relationshipss?

File: website/docs/d/l3_outside.html.markdown [hunk 15]
```diff
@@ -10,25 +10,40 @@ description: |-
...
+* `description`- (Optional) Description of the L3 Outside object.
+* `annotation` - (Optional) Annotation of the L3 Outside object.
+* `enforce_rtctrl` - (Optional) Enforce route control type of the L3 Outside object. 
+* `name_alias` - (Optional) Name alias of the L3 Outside object.
+* `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the L3 Outside object.
+* `relation_l3ext_rs_dampening_pol` - (Optional) Relation to Route Control Profile for Dampening Policies (class rtctrlProfile). Cardinality - N_TO_M. Type - Block.
+  * tn_rtctrl_profile_name - (Deprecated) Name of the Route Control Profile for Dampening Policies.
+  * tn_rtctrl_profile_dn - (Optional) Distinguished name of the Route Control Profile for Dampening Policies.
+  * af - (Optional) Address Family of the Dampening Policies.
+* `relation_l3ext_rs_ectx` - (Optional) Relation to VRF (class fvCtx).
```
Comment (akinross) [inline website/docs/d/l3_outside.html.markdown:47]: should we include cardinality and extra description for these as is done in the resource?

**Review by lhercot — APPROVED** [review]
LGTM

**Review by akinross — APPROVED** [review]
LGTM

**Pull request merged**